OpenSim/Simulation/SimbodyEngine/Joint.cpp

```diff
@@ -266,43 +266,56 @@ bool Joint::isCoordinateUsed(const Coordinate& aCoordinate) const
+    // (which can occur only if the Frame is a PhysicalOffsetFrame).
+    auto findIndexInFramesList = [&](const PhysicalFrame& frame) -> int
+    {
+        const PhysicalOffsetFrame* pof =
+            dynamic_cast<const PhysicalOffsetFrame*>(&frame);
+        if (pof == nullptr)
+            return -1;
+        return getProperty_frames().findIndex(*pof);
+    };
+    const int parentIndex = findIndexInFramesList(getParentFrame());
+    const int childIndex  = findIndexInFramesList(getChildFrame());
+
+    if (parentIndex < 0 && childIndex < 0)  // No scaling to do.
+        return;
+
+    // If parent and/or child is owned by the Joint, search the scaleSet *once*
+    // to find the associated scale factors.
+    const std::string& parentName = getParentFrame().findBaseFrame().getName();
+    const std::string& childName  =  getChildFrame().findBaseFrame().getName();
     SimTK::Vec3 parentFactors(1.0);
     SimTK::Vec3 childFactors(1.0);
+    bool seekingParent = (parentIndex >= 0);
+    bool seekingChild  = (childIndex  >= 0);
 
-    // Find the factors associated with the PhysicalFrames this Joint connects
-    const string& parentName = getParentFrame().findBaseFrame().getName();
-    const string& childName = getChildFrame().findBaseFrame().getName();
-    // Get scale factors
-    bool found_p = false;
-    bool found_b = false;
-    for (int i = 0; i < scaleSet.getSize(); i++) {
+    for (int i=0; i<scaleSet.getSize(); ++i) {
         Scale& scale = scaleSet.get(i);
-        if (!found_p && (scale.getSegmentName() == parentName)) {
+        // TODO: The scaling infrastructure must be updated now that bodies can
+        //       have the same name (provided their absolute paths are unique).
+        if (seekingParent && (scale.getSegmentName() == parentName)) {
             scale.getScaleFactors(parentFactors);
-            found_p = true;
+            seekingParent = false;
         }
-        if (!found_b && (scale.getSegmentName() == childName)) {
+        if (seekingChild && (scale.getSegmentName() == childName)) {
             scale.getScaleFactors(childFactors);
-            found_b = true;
+            seekingChild = false;
         }
-        if (found_p && found_b)
+        if (!seekingParent && !seekingChild)
             break;
     }
 
-    // if the frame is owned by this Joint scale it,
-    // otherwise let the owner of the frame decide.
-    int found = getProperty_frames().findIndex(getParentFrame());
-    if (found >= 0) {
-        PhysicalOffsetFrame& offset
-            = SimTK_DYNAMIC_CAST_DEBUG<PhysicalOffsetFrame&>(upd_frames(found));
-        offset.scale(parentFactors);
-    }
-    found = getProperty_frames().findIndex(getChildFrame());
-    if (found >= 0) {
-        PhysicalOffsetFrame& offset
-            = SimTK_DYNAMIC_CAST_DEBUG<PhysicalOffsetFrame&>(upd_frames(found));
-        offset.scale(childFactors);
-    }
+    OPENSIM_THROW_IF(seekingParent || seekingChild, Exception,
```

Comment:
I believe the current expectation is that the `ScaleSet` defines what is to be scaled, not the existence of the `PhyscialOffsetFrame`. In other words, if there is no `ScaleFactor` for this offset's base frame then it just doesn't get scaled. If you purposefully remove scale factors for a `Body` from the `ScaleSet` would should scaling throw?

Reply:
I think ScaleSet should contain exactly one entry for each Body (even if that entry is `Vec3(1)`). Searching through a ScaleSet and scaling by `Vec3(1)` by default if no entry is found sounds dangerous (e.g., a Body name could be misspelled and skipped over silently). ScaleSet's `finalizeFromProperties()` should ensure there is exactly one entry for each Body. If that is not currently the case, I think it should be made so. Having one entry per Body will also improve the search time (will always require `n(n+1)/2` string comparisons for `n` bodies; the worst-case cost is `n^2` otherwise).

Reply:
> Searching through a ScaleSet and scaling by Vec3(1) by default if no entry is found sounds like a bad design (e.g., a Body name could be misspelled and skipped over without noticing).

That is not exactly what I proposed. If a `Body` is not listed it should not be scaled - not scaled by Vec3(1).  A `ScaleFactor` with an illegitimately named `Body` should be made to throw. Your approach would require the user to know about every `Body` in the `Model` in order to scale it. If I embed bodies in a Device model that implements its own scale say based on distances between anchor points, it seems far too burdensome for the consumer of that Device to have to include its bodies in order for the scaling to work.